core/storage/trie/polkadot_trie/polkadot_trie_cursor_impl.cpp

```diff
@@ -101,159 +110,149 @@ namespace kagome::storage::trie {
-      current = new_current;
-    } else {
-      current = trie_.getRoot();
-    }
-    last_visited_child_ = followed_path;
-    nibbles = KeyNibbles{Buffer{left_nibbles}};
-
-    if (nibbles.empty()) {
-      if (current->value.has_value()) {
-        current_ = current;
-        return true;
-      } else {
-        // no value - certainly not a leaf
-        auto current_as_branch = std::dynamic_pointer_cast<BranchNode>(current);
-        OUTCOME_TRY(desc, seekChildWithValue(current_as_branch));
-        if (desc.has_value()) {
-          current_ = desc.value();
-        }
-        return desc.has_value();
-      }
-    } else if (not nibbles.empty()) {
-      if (current->getTrieType() == NodeType::Leaf) {
-        if (nibbles < current->key_nibbles) {
-          current_ = current;
-          return true;
-        } else {
-          current_ = nullptr;
-          return false;
-        }
-      }
-      auto current_as_branch = std::dynamic_pointer_cast<BranchNode>(current);
-      OUTCOME_TRY(descendant,
-                  seekChildWithValueAfterIdx(current_as_branch, nibbles[0]));
-      if (descendant.has_value()) {
-        current_ = descendant.value();
-        return true;
-      } else {
-        current_ = nullptr;
-        return false;
-      }
-    } else {
-      // nibbles are equal
-      current_ = current;
-      return true;
-    }
-  }
-
-  outcome::result<std::list<PolkadotTrieCursorImpl::TriePathEntry>>
-  PolkadotTrieCursorImpl::followNibbles(
-      NodePtr node, gsl::span<const uint8_t> &left_nibbles) const {
-    if (node == nullptr) {
-      return {{}};
-    }
-    std::list<TriePathEntry> followed_path;
-    NodePtr current = node;
-    while (not left_nibbles.empty()) {
-      auto &&[left_nibbles_mismatch, current_nibbles_mismatch] =
+    while (true) {
+      auto [left_mismatch, current_mismatch] =
           std::mismatch(left_nibbles.begin(),
                         left_nibbles.end(),
                         current->key_nibbles.begin(),
                         current->key_nibbles.end());
-      size_t common_length = left_nibbles_mismatch - left_nibbles.begin();
-      switch (current->getTrieType()) {
-        case NodeType::BranchEmptyValue:
-        case NodeType::BranchWithValue: {
-          // ran out of left_nibbles
-          if (current->key_nibbles == left_nibbles or left_nibbles.empty()
-              or static_cast<size_t>(left_nibbles.size()) < current->key_nibbles.size()) {
-            left_nibbles = left_nibbles.subspan(common_length);
-            return followed_path;
+      // parts of every sequence within their common length (minimum of their
+      // lenghts) are equal
+      bool part_equal = left_mismatch == left_nibbles.end()
+                        or current_mismatch == current->key_nibbles.end();
+      // if current choice is lexicographically less or equal to the left part
+      // of the sought key, we just take the closest node with value
+      bool lessOrEq =
```

Comment:
It violates of name convention

Reply:
Oh, that's why it looks so ugly, now I see.